src/components/GiftRankingHeader.tsx

```diff
@@ -0,0 +1,204 @@
+import { css } from "@emotion/react";
+import type { Theme } from "@emotion/react";
+import { useTheme } from "@emotion/react";
+import { useLocation, useNavigate } from "react-router-dom";
+import { useEffect } from "react";
+
+type GiftRankingHeaderProps = {
+  target: string;
+  setTarget: React.Dispatch<React.SetStateAction<string>>;
+  rankType: string;
+  setRankType: React.Dispatch<React.SetStateAction<string>>;
+};
+
+const GiftRankingHeader: React.FC<GiftRankingHeaderProps> = ({
+  target,
+  setTarget,
+  rankType,
+  setRankType,
+}) => {
+  type RANK_TYPE = "MANY_WISH" | "MANY_RECEIVE" | "MANY_WISH_RECEIVE";
+
+  const navigate = useNavigate();
+  const theme = useTheme();
+  const location = useLocation();
+
+  useEffect(() => {
+    const searchParams = new URLSearchParams(location.search);
+    const initTarget = searchParams.get("target") || "ALL";
+    const initRank = (searchParams.get("rankType") as RANK_TYPE) || "MANY_WISH";
+    setTarget(initTarget);
+    setRankType(initRank);
+    const params = new URLSearchParams(location.search);
+    params.set("target", initTarget);
+    params.set("rankType", initRank);
+    navigate(`${location.pathname}?${params.toString()}`);
+  }, [location.pathname, location.search, navigate, setRankType, setTarget]);
+
+  const handleTargetClick = (newTarget: string) => {
+    setTarget(newTarget);
+    const params = new URLSearchParams(location.search);
+    params.set("target", newTarget);
+    params.set("rankType", rankType);
+    navigate(`${location.pathname}?${params.toString()}`);
+  };
+
+  const handleRankClick = (newRank: string) => {
+    setRankType(newRank);
+    const params = new URLSearchParams(location.search);
+    params.set("target", target);
+    params.set("rankType", newRank);
+    navigate(`${location.pathname}?${params.toString()}`);
+  };
```

Comment:
useSearchParams 훅을 사용하면, 코드를 비교적 간단하게 작성할 수 있어요!

```suggestion
  const [searchParams, setSearchParams] = useSearchParams();

  useEffect(() => {
    const initTarget = searchParams.get("target") || "ALL";
    const initRank = (searchParams.get("rankType") as RANK_TYPE) || "MANY_WISH";
    setTarget(initTarget);
    setRankType(initRank);
  }, [location.pathname, location.search, navigate, setRankType, setTarget, searchParams]);

  const handleTargetClick = (newTarget: string) => {
    setTarget(newTarget);
    searchParams.set("target", newTarget);
    searchParams.set("rankType", rankType);
    setSearchParams(searchParams);
  };

  const handleRankClick = (newRank: string) => {
    setRankType(newRank);
    searchParams.set("target", target);
    searchParams.set("rankType", newRank);
    setSearchParams(searchParams);
  };
```

Reply:
아하 이런 방법도 있었네요 useSearchParam hook을 써서 searchParam을 상태로 관리하면 변경될 때마다 navigate로 이동하는 느낌이 아니라 경로를 계속 업데이트해서 반영하는 느낌이네요.
이것도 반영해서 다시 리뷰 요청하겠습니다~